File: leetcode/medium/377-comb-sum-4/tab-dp.cpp
/**
 * @file tab-dp.cpp
 * @author Vortexx2
 * @brief Problem 377 - Combination Sum 4
 * @approach Bottom up DP tabulation approach
 * @date 02-10-2022
 * 
 * Runtime - 5 ms O(nums * target)
 * Memory Usage - 6.6 MB O(target)
 */
#include <iostream>
#include <map>
#include <vector>

using namespace std;

class Solution {
 public:
  int combinationSum4(vector<int>& nums, int target) {
    if (!target) return 1;

    int n = nums.size();
    vector<int> dp(target + 1, 0);
    dp[0] = 1;

    for (int subTarget = 1; subTarget <= target; subTarget++) {
      int ways = 0;
      for (const int& num : nums) {
        if (num == subTarget) ways += 1;

        if (num < subTarget) ways += dp[subTarget - num];
      }

      dp[subTarget] = ways;
    }

    return dp[target];
  }
};

File: leetcode/medium/377-comb-sum-4/tab-dp.py
"""
  @author Vortexx2
  Problem 377 - Combination Sum 4
  @approach Using a bottom up tabulation DP approach

  Runtime - 32 ms O(nums * target)
  Memory Usage - 13.9 MB O(target)
"""
from typing import List


class Solution:
  def combinationSum4(self, nums: List[int], target: int) -> int:

    if not target:
      return 1

    n = len(nums)
    dp = [0] * (target + 1)
    dp[0] = 1

    for subAmount in range(1, target + 1):

      ways = 0

      for num in nums:
        if num == subAmount:
          ways += 1

        if num < subAmount:
          ways += dp[subAmount - num]

      dp[subAmount] = ways

    return dp[target]
